### ISSUE TITLE
DeepSpeech Engine

### ISSUE THREAD

**Issue by multimeric** [dictation-toolbox/dragonfly#139]
Hi all, I've used [DeepSpeech](https://github.com/mozilla/DeepSpeech) before to pretty good effect. It's a deep-learning based recognition tool, which could give us advantages by being:
* Cross platform
* GPU enabled
* Models provided
* High accuracy
* Streamable

It also has a published python package, and is regularly updated. Would there be any interest in an engine that uses DeepSpeech? Does anyone have any pointers for writing a new engine?

**Comment by Danesprite** [comment]
Thanks for posting this. DeepSpeech has come up before in the [Gitter channel](https://gitter.im/dictation-toolbox/dragonfly). From memory, I think there were some questions raised about how accurate models would be given that Dragonfly grammars are typically command-based, rather than free-form dictation. Having said that, it would be nice to try it.

It is my understanding that you need to set up voice activity detection (VAD) yourself. Since we now have the [Kaldi engine backend](https://dragonfly2.readthedocs.io/en/latest/kaldi_engine.html), which also had this requirement, it is probably possible to re-use the VAD and audio code used in that engine.

As for writing a new engine, there is a section on this in dragonfly's [contributing file](https://github.com/dictation-toolbox/dragonfly/blob/master/CONTRIBUTING.rst#additional-sr-engine-backends). It could probably be improved though.

@daanzu What are your thoughts on all this?

**Comment by daanzu** [comment]
I have done some work with DeepSpeech, and considered implementing a dragonfly backend for it, but I decided Kaldi was a more appropriate choice.

In my opinion, a huge part of dragonfly are the strict command grammars, and DeepSpeech does not currently support anything other than free form dictation. Implementing such support is possible but would be a significant undertaking in the internals of DeepSpeech. Without this, the accuracy of commands is likely to be fairly poor. Also, its general dictation accuracy and performance seemed underwhelming compared to what Kaldi is capable of.

Kaldi is much more mature and flexible, and supports strict grammar parsing. With my active grammar extension, it fully supports all of dragonfly. Have you tried the Kaldi backend? I think it checks all of your boxes (except GPU decoding which it doesn't need to achieve high performance), and I would be surprised if DeepSpeech could improves upon it.

Nevertheless, implementing a partial DeepSpeech backend would not be overly difficult, using the Kaldi backend as a template.

**Comment by TMiguelT** [comment]
My concern was mostly with the accuracy of free text. I was concerned that, outside of fixed commands, Kaldi would be inaccurate for dictating actual sentences, whereas DeepSpeech is ideal for that use-case. I'm interested in voice coding, and I'm not sure how effective that would be without high accuracy free text, in addition to command grammars.

**Comment by daanzu** [comment]
I would say free dictation is the ideal case for DeepSpeech, but not necessarily vice versa. The "state of the art" is highly debatable, but Kaldi is certainly in the top tier: https://github.com/syhw/wer_are_we (I actually haven't run that benchmark on my current Kaldi model but I intend to). Note that the Mozilla streaming implementation is less accurate than the older bidirectional implementation, which was itself less accurate than the original Baidu implementation results in the paper. Nevertheless, it would actually be quite easy to modify the Kaldi backend's cloud dictation option to use any local python package like DeepSpeech, if there is interest.

**Comment by Danesprite** [comment]
Modifying the Kaldi backend's cloud dictation option to use any local Python package sounds like a good idea to me.

**Comment by TMiguelT** [comment]
Yep, that sounds like an excellent compromise. Would that involve a kaldi PR? Or a PR here?

**Comment by daanzu** [comment]
Most of the changes will be in Kaldi active grammar, but the options to the dragonfly backend may need to be adjusted as well.

However, I just realized that DeepSpeech requires python3, even for inference, correct? Kaldi active grammar currently only runs under python2, though I hope to get python3 support completed soon.

**Comment by TMiguelT** [comment]
Ah, it quite possibly is Python 3. In any case, it would be a better use of time finishing Python 3 support in active grammar than trying to hack Deepspeech into working with Python 2.

**Comment by TMiguelT** [comment]
Am I correct in saying that active grammar only currently works with the (paid) Google cloud dictation service? So DeepSpeech sound like it could be a huge improvement on that front.

**Comment by daanzu** [comment]
The cloud dictation feature only works with google currently, though it should not be hard to add others depending on their interface. However, the cloud dictation feature is just optional. By default, full dictation is supported locally by Kaldi.

**Comment by LexiconCode** [comment]
Some changes have been made to make the in  [kaldi-active-grammar
v1.0.0](https://github.com/daanzu/kaldi-active-grammar/releases/tag/v1.0.0) to make it easier to add other alternatives like DeepSpeech for dictation!

**Comment by TMiguelT** [comment]
Excellent! Where can other cloud dictation engines be hooked in?

**Comment by daanzu** [comment]
Unfortunately, I don't have a "pluggable" interface for it yet, but the basic infrastructure is there. I'll try to improve it when I can.

**Comment by daanzu** [comment]
Okay, you made it bother me 😉, so I wrote a pluggable interface. See #184. It should be trivial to hook up deepspeech. Let me know if you have trouble.

**Comment by TMiguelT** [comment]
Wow, amazing work. I had a look through the PR but I can't find mention of the interface itself. Would you mind pointing me to the right code?

**Comment by daanzu** [comment]
Oops, I'll post documentation for the interface.

**Comment by daanzu** [comment]
Set `alternative_dictation` to a ``callable`` object. The callable must accept at
  least one argument (for the audio data) and any keyword arguments. The
  audio data is passed in standard Linear16 (``int``) PCM encoding.

**Comment by daanzu** [comment]
Oh, the callable should return the recognized text.

**Comment by tbenst** [comment]
Just wanted to post this resource for [real-time streaming using microphone from python or web browser with DeepSpeech](https://github.com/mozilla/DeepSpeech-examples/tree/r0.6/mic_vad_streaming).

Another approach is to follow the example set by Common Voice and actually obtain a corpus for voice coders. This could be bootstrapped for e.g. the Casper vocabulary by recording audio from users + capturing a Dragon/Kaldi interpretation that is then manually corrected. This data could then be used for transfer learning--use pretrained model, and retrain on our specialized corpus.

**Comment by daanzu** [comment]
Thanks for the interest! Though actually, I wrote that module for DeepSpeech 😀.

As you referred to in Gitter, handling the command grammars is a difficulty. I have worked some with DeepSpeech, and considered implementing a decoder for it that would support command grammars, but I am not sure its performance justifies that currently. So I chose to work on Kaldi instead, which I expect to perform better in this use case, especially with a lesser amount of training data for commands.

I would love to collect a corpus of data of command usage to use for training! I am currently working on training better Kaldi models, especially for commands, but there is a dearth of data available for commands especially.

FYI, I recently added support for retaining recognition metadata and/or audio in dragonfly (for natlink/Kaldi/WSR backends). I have used it in part to create a custom model for myself with fantastic results. There are various ways of trying to handle the issue of the correctness of the recognitions, which we could discuss.
